### PR TITLE
[org-node] add Client::track_peers process for tracking and fetching remote peers

### LINKED ISSUE #46: [org-node] track project peers
The `org-node` currently tracks projects, but not its peers. When the `org-node` discovers a known project with an unknown remote (peer), the remote should be tracked and its version of the project fetched.

### PR DESCRIPTION
**Pull request by Ryanmtate**
Resolves #46

### CODE REVIEW

File: org-node/src/client.rs [hunk 1]
```diff
@@ -277,70 +523,24 @@ impl Client {
...
-                            // maintainer.
-                            match Client::set_head(&urn, &remote, &branch, paths) {
-                                Err(Error::SetHead(err)) => {
-                                    tracing::error!(target: "org-node", "Error setting head: {}", err);
-                                }
-                                Err(err) => return Err(err),
-                                Ok(()) => {}
-                            }
-                            Client::sign_refs(urn, api).await?;
+                if let Some(peer) = peers.next().await {
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply org-node/src/client.rs:520]: Hmm I don't think that's right: the while loop is there because we are attempting peers one by one until the first one succeeds to replicate. With an `if`, we would only be trying the first peer we find.

[the remark this answers is not in the document]
Comment (Ryanmtate) [reply org-node/src/client.rs]: 3b77898 reverts to using while loop; I think what I was missing originally was continuing on an error for the match arm, such that the loop does not return an early early, breaking the loop.

File: org-node/src/client.rs [hunk 2]
```diff
@@ -125,6 +141,19 @@ pub struct Head {
...
 }
 
+/// Project peer contains details used to track the peer for an interested project Urn.
+/// this struct is primarily used by `ProjectPeerTransmitter`
+struct ProjectPeer {
+    pub api: Peer<Signer>,
+    pub peer_info: PeerInfo<std::net::SocketAddr>,
+    pub paths: Paths,
+    pub urn: Urn,
+}
```
Comment (cloudhead) [inline org-node/src/client.rs:151]: Do the fields need to be public? The struct isn't.

Comment (cloudhead) [reply org-node/src/client.rs:151]: I would also find a better name for this that communicates its intent, eg. `TrackPeer`, since it's basically a command, right?

File: org-node/src/client.rs [hunk 3]
```diff
@@ -235,11 +285,200 @@ impl Client {
...
+            let api = rx.0.api;
+            let urn = rx.0.urn;
+            let peer_info = rx.0.peer_info;
+            let paths = rx.0.paths;
+
+            let ack = rx.1;
+
+            match Client::track_project(&api, &urn, &peer_info).await {
+                Err(err) => {
+                    tracing::error!(target: "org-node", "Error tracking {}: {}", urn, err);
```
Comment (cloudhead) [inline org-node/src/client.rs:302]: We don't return any error back, so how does the sender know to retry?

File: org-node/src/client.rs [hunk 4]
```diff
@@ -235,11 +285,200 @@ impl Client {
...
+                Ok(Err(e)) => {
+                    tracing::error!(target: "org-node", "Project peer returned error: {}", e);
+                    Err(e)
+                }
+                Err(e) => {
+                    tracing::error!(target: "org-node", "Error project peer receiver failed: {}", e);
+                    Err(Error::Recv(e))
+                }
+                Ok(res) => res,
+            },
```
Comment (cloudhead) [inline org-node/src/client.rs:414]: Any way to avoid these nested result matches? It's a bit hard to follow.

File: org-node/src/client.rs [hunk 5]
```diff
@@ -231,6 +247,53 @@ impl Client {
         }
     }
 
+    /// Track peers that are found by the node that are interested in shared projects;
+    /// This process is spawned within `Client::run` and is run indefinitely.
+    async fn track_peers(node: Peer<Signer>, known_peers: Vec<PeerId>, allow_unknown_peers: bool) {
+        tracing::info!(target: "org-node", "Spawning Peer Tracking Listener");
+        let events = node.subscribe();
+        pin_mut!(events);
```
[the remark this answers is not in the document]
Comment (cloudhead) [reply org-node/src/client.rs:433]: Well, it's a macro that shadows a variable with another variable, from my understanding, ie. it has some effects on the surrounding code. It was created to patch the deficiencies of the async/await feature of the language.. I can try to see if we can get around it, I just find it very hacky.

File: org-node/src/client.rs [hunk 6]
```diff
@@ -282,69 +521,30 @@ impl Client {
                 let mut peers = api.providers(urn.clone(), timeout);
 
                 while let Some(peer) = peers.next().await {
-                    match Client::track_project(api, &urn, &peer).await {
-                        Err(err) => {
-                            tracing::error!(target: "org-node", "Error tracking {}: {}", urn, err);
+                    // Send notification to track peers process;
+                    let response = match Client::track_peer_handler(
```
Comment (cloudhead) [inline org-node/src/client.rs:525]: Maybe move this out to make the code a little easier to read:
```rust
let result = Client::track_peer_handler(...).await;
let response = match result { ...
```

File: org-node/src/client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ryanmtate) [reply org-node/src/client.rs:433]: Hmm, I thought it is to pin the memory to the stack without needing to write your own unsafe code.

File: org-node/src/client.rs (hunk 2 above)
Comment (Ryanmtate) [reply org-node/src/client.rs:151]: 6e416d0 renames, 5dfbfd6 removes `pub` fields.

File: org-node/src/client.rs [hunk 7]
```diff
@@ -235,11 +285,200 @@ impl Client {
...
 
+    /// Track peers that are found by the node that are interested in shared projects;
+    /// This process is spawned within `Client::run` and is run indefinitely.
+    async fn track_peers(mut peer_rx: mpsc::Receiver<ProjectPeerTransmitter>) {
+        tracing::info!(target: "org-node", "Spawning Track Peers Listener");
+        while let Some(rx) = peer_rx.recv().await {
+            let api = rx.0.api;
+            let urn = rx.0.urn;
+            let peer_info = rx.0.peer_info;
+            let paths = rx.0.paths;
```
[the remark this answers is not in the document]
Comment (Ryanmtate) [reply org-node/src/client.rs]: b80afa3 destructures parameters within the if let binding.

File: org-node/src/client.rs (hunk 3 above)
Comment (Ryanmtate) [reply org-node/src/client.rs:304]: Thanks, nice catch! 5dfbfd6 adds.

File: org-node/src/client.rs (hunk 4 above)
Comment (Ryanmtate) [reply org-node/src/client.rs]: 5dfbfd6 removes nested match arms

File: org-node/src/client.rs (hunk 6 above)
Comment (Ryanmtate) [reply org-node/src/client.rs]: 5dfbfd6 fixes.

File: org-node/src/client.rs [hunk 8]
```diff
@@ -282,69 +513,33 @@ impl Client {
...
+                    )
+                    .await;
+
+                    // Handle the track peer response
+                    let response = match result {
+                        Ok(response) => response,
+                        Err(e) => {
+                            // QUESTION: Should we reply with an error here or continue?;
+                            tracing::error!(target: "org-node", "Error tracking peers: {:?}", e);
+                            continue;
```
Comment (Ryanmtate) [inline org-node/src/client.rs:534]: Should we be returning an error here instead of continuing?

```
reply
.send(Err(TrackProjectError::<REASON>))
.map_err(|_| Error::Reply("TrackProject".to_string()))
```

File: org-node/src/client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (cloudhead) [reply org-node/src/client.rs:425]: it is, but it does so by shadowing the var with another var, see: https://docs.rs/pin-utils/0.1.0/src/pin_utils/stack_pin.rs.html#14-25

File: org-node/src/client.rs (hunk 8 above)
Comment (cloudhead) [reply org-node/src/client.rs:534]: No, we return a `NotFound` error if none of the peers we try succeed. It's okay if some of the attempts fail, we just care if one succeeds.

Comment (Ryanmtate) [reply org-node/src/client.rs:533]: eaa3a24 removes question comment.

File: org-node/src/client.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (Ryanmtate) [reply org-node/src/client.rs]: 9ba4f7a removes and uses `.for_each`

**Review by cloudhead — APPROVED** [review]
